File: octopoints_flutter/lib/service/impl/TeamServiceImpl.dart
import 'package:octopoints_flutter/db/DbSingleton.dart';
import 'package:octopoints_flutter/db/dao/TeamDao.dart';
import 'package:octopoints_flutter/db/dao/TeamUserRelationDao.dart';
import 'package:octopoints_flutter/db/dao/UserDao.dart';
import 'package:octopoints_flutter/db/entity/TeamUserRelation.dart';
import 'package:octopoints_flutter/service/TeamService.dart';
import 'package:octopoints_flutter/service/model/User.dart';
import 'package:octopoints_flutter/service/model/Team.dart';
import 'package:octopoints_flutter/service/mapper/TeamMapper.dart';
import 'package:octopoints_flutter/service/mapper/UserMapper.dart';
import 'package:octopoints_flutter/service/service.dart';

class TeamServiceImpl implements TeamService {
  final TeamDao _teamDao = DbSingleton().db.teamDao;
  final TeamUserRelationDao _relationDao = DbSingleton().db.teamUserRelationDao;

  @override
  Future<List<int>> addTeammates(int teamId, List<User> users) {
    List<TeamUserRelation> relations = [];
    for (User element in users) {
      relations.add(TeamUserRelation(teamId, element.id));
    }
    return _relationDao.addTeammates(relations);
  }

  @override
  Future<Team> createTeam(Team team) async {
    int id = await _teamDao.create(team.toTeamEntity());
    return team.setId(id);
  }

  @override
  Future<int> deleteTeam(Team team) {
    return _teamDao.remove(team.toTeamEntity());
  }

  Future<List<User>> _getTeammates(int teamId) async {
    return (await _relationDao.getTeammates(teamId))
        .map((e) => e.toUserModel())
        .toList();
  }

  @override
  Future<List<Team>> getTeamsByMatchId(int id) async {
    List<Team> teams = (await _teamDao.getTeamsByMatchId(id))
        .map((e) => e.toTeamModel())
        .toList();
    for (Team team in teams) {
      team.users = await _getTeammates(team.id);
    }
    return teams;
  }

  @override
  Future<int> removeTeammates(int teamId, int userId) {
    return _relationDao.remove(TeamUserRelation(teamId, userId));
  }

  @override
  Future<int> updateTeams(List<Team> teams) {
    return _teamDao.modify(teams.map((e) => e.toTeamEntity()).toList());
  }

  @override
  Future<Map<User, bool>> getAvailableTeammates(Team team) async {
    List<User> users = await DBService.userService.getUsers();
    for (Team item in (await getTeamsByMatchId(team.matchId))) {
      for(User user in item.users){
        users.removeWhere((element) => element.id == user.id);
      }
      
    }
    Map<User, bool> availableUser =
        { for (var e in users) e : false };
    availableUser.addAll({ for (var e in team.users) e : true });
    return availableUser;
  }
}

File: octopoints-kotlin/app/src/main/java/dev/tpavani/octopoints/service/impl/TeamServiceImpl.kt
package dev.tpavani.octopoints.service.impl

import dev.tpavani.octopoints.db.DbSingleton
import dev.tpavani.octopoints.db.entity.TeamUserRelationEntity
import dev.tpavani.octopoints.service.TeamService
import dev.tpavani.octopoints.service.model.Team
import dev.tpavani.octopoints.service.model.User
import dev.tpavani.octopoints.service.utils.ModelMapper
import java.util.*

class TeamServiceImpl: TeamService {
    override suspend fun createTeam(team: Team): Team {
        val id: Int = DbSingleton.getDb().teamDao().create(ModelMapper.toTeamEntity(team))
        return team.setId(id)
    }

    override suspend fun updateTeams(teams: List<Team>) {
        DbSingleton.getDb().teamDao().modify(teams.map { e-> ModelMapper.toTeamEntity(e) })
    }

    override suspend fun deleteTeam(team: Team) {
        DbSingleton.getDb().teamDao().remove(ModelMapper.toTeamEntity(team))
    }

    private suspend fun getTeammates(teamId: Int): List<User>{
        return DbSingleton.getDb().teamUserRelationDao().getTeammates(teamId).map { e-> ModelMapper.toUserModel(e) }
    }

    override suspend fun getTeamsByMatchId(id: Int): List<Team> {
        var teams: MutableList<Team> = mutableListOf()
        for(teamEntity in DbSingleton.getDb().teamDao().getTeamsByMatchId(id)){
            var team: Team = ModelMapper.toTeamModel(teamEntity)
            team.users.addAll(getTeammates(team.getId()))
            teams.add(team)
        }
        return teams
    }

    override suspend fun addTeammates(teamId: Int, users: List<User>) {
        var relations: MutableList<TeamUserRelationEntity> = mutableListOf()
        for (user in users){
            relations.add(TeamUserRelationEntity(teamId, user.getId()))
        }
        DbSingleton.getDb().teamUserRelationDao().addTeammates(relations)
    }

    override suspend fun updateTeammates(team: Team) {
        DbSingleton.getDb().teamUserRelationDao().deleteTeammates(team.getId())
        addTeammates(team.getId(), team.users)
    }

    override suspend fun getAvailableTeammates(team: Team): Map<User, Boolean> {
        var users: MutableList<User> = mutableListOf()
        users.addAll(DbSingleton.getDb().userDao().getAllUsers().map { e-> ModelMapper.toUserModel(e) })
        for(team in getTeamsByMatchId(team.getMatchId())){
            users.removeAll { user-> team.users.map { e-> e.getId() }.contains(user.getId()) }
        }
        var map: SortedMap<User, Boolean> = sortedMapOf(comparator = compareBy{it.getUsername()})

        map.putAll(users.map { e-> e to false }.toMap())
        map.putAll(team.users.map { e-> e to true }.toMap())

        return map
    }
}